### DISCUSSION TITLE
auth.getUser() returns null despite having a session

### DISCUSSION THREAD

**Discussion by Ryurcia** [supabase/supabase#22533] [Questions] [answered]
I am currently using the `@supabase/ssr` package for my auth. Followed the docs and everything. In most my server components, I need to fetch the current users UID, so I use the getUser() method (which was recommended) to retrieved the UID.  It would sometimes return the user object and sometimes not? Any reason why. 

Server Client:
```
export function createClient() {
  const cookieStore = cookies();

  return createServerClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL!,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY!,
    {
      cookies: {
        get(name: string) {
          return cookieStore.get(name)?.value;
        },
        set(name: string, value: string, options: CookieOptions) {
          try {
            cookieStore.set({name, value, ...options});
          } catch (error) {
          }
        },
        remove(name: string, options: CookieOptions) {
          try {
            cookieStore.set({name, value: '', ...options});
          } catch (error) {
          }
        },
      },
    }
  );
}
```
Profile Page:
```
const page = async () => {
  const supabase = createClient();
  const{data:{user},error} = await supabase.auth.getUser();

  if(!user) throw 'No user retrieved';
  if(error) throw error.message

...rest of my code
}
```

**Comment by rookledookle** [comment]
Running into the same issue with /ssr package- seems like quite a lot of people are too

**Reply by Ryurcia** [reply]
Its working now for me, but I haven't changed any of my code. Still dont know why it was doing that.

**Comment by Cainor** [comment]
Same issue here. Everything is exactly as the doc says, I'm able to login and able to register, but the `getUser()` function always return null.
Could be something related to cookies.

Note that I'm able to see the login logs in my Supabase logs:
![image](https://github.com/supabase/supabase/assets/36992322/662a9487-cdaa-406c-bd08-5b78ae18f9f2)

**Reply by Cainor** [reply]
I think my problem is that the user does not have cookies at all. After logging in, I've checked the browser cookies, and there were none.
Now, time to learn how to set cookies correctly.

**Reply by Cainor** [reply, marked as answer]
I've solved my issue.

Here is where I tried to be a good programmer without understanding what is going on. First of all I'm using Next.js and following this doc: https://supabase.com/docs/guides/auth/server-side/nextjs

In the "actions.ts" file, I saw that the doc is declaring `const supabase = createClient()` in every function, and here I thought: "hmm, this is not efficient" so this is what I did: 
```node
import { createClient } from "~/utils/supabase/server";

const supabase = createClient();

export async function login(formData: FormData) {
...
}
export async function signup(formData: FormData) {
...
}

```

This, somehow has the login function working properly except that it never stores cookies.

Fixing it like the doc:
```node
import { createClient } from "~/utils/supabase/server";

export async function login(formData: FormData) {
const supabase = createClient();
...
}
export async function signup(formData: FormData) {
const supabase = createClient();
...
}
```

Now this works fine and as intended. If anyone knows what is going on here, please let me know.

Thank you all.

**Reply by Ryurcia** [reply]
That was my issue too. I had only one client created at the top of my file that I reused for every action. I then decided to create a client under every action which got it working. It definitely had something to do with the cookies being set.

**Reply by Cainor** [reply]
If possible, please mark the discussion as “Answered”. Also, if you think the way I documented the solution is fair to mark it as the solution, I’d love to get a GitHub badge on my profile 😁
